python/tvm/relax/transform/op_legalizer.py

```diff
@@ -0,0 +1,559 @@
+                    a_dim_is_one = isinstance(a_dim, tvm.tir.IntImm) and a_dim == 1
+                    b_dim_is_one = isinstance(b_dim, tvm.tir.IntImm) and b_dim == 1
+                    a_indices.append(0 if a_dim_is_one else idx_spatial[i])
+                    b_indices.append(0 if b_dim_is_one else idx_spatial[i])
+                if not a_prepended:
+                    a_indices.append(idx_spatial[-2 + b_appended])
+                a_indices.append(idx_reduce)
+                b_indices.append(idx_reduce)
+                if not b_appended:
+                    b_indices.append(idx_spatial[-1])
+
+                dtype = call.attrs.out_dtype
+                if dtype != "":
+                    return a(*a_indices).astype(dtype) * b(*b_indices).astype(dtype)
+                else:
+                    return a(*a_indices) * b(*b_indices)
+
+            return te.sum(multiply_compute(k), axis=k)
+
+        return te.compute(
+            output_shape,
+            lambda *idx: matmul_compute(*idx),  # pylint: disable=unnecessary-lambda
+            name="matmul",
+        )
+
+    return bb.call_te(te_matmul, call.args[0], call.args[1], primfunc_name_hint="matmul")
+
+
+def _nn_softmax(bb: BlockBuilder, call: Call):
+    return bb.call_te(topi.nn.softmax, call.args[0], call.attrs.axis)
+
+
+def _flatten(bb: BlockBuilder, call: Call):
+    return bb.call_te(topi.nn.flatten, call.args[0])
+
+
+def _nn_adaptive_max_pool2d(bb: BlockBuilder, call: Call):
+    if call.attrs.out_layout != call.attrs.layout:
+        logging.info(
+            "TOPI adaptive_max_pool2d does not support different input-output "
+            "layouts, and thus cannot be legalized by TOPI"
+        )
+        return call
+
+    if call.attrs.output_size is None:
+        # Todo(ruihang): seems that TOPI adaptive_avg_pool cannot handle the
+        # cases where output_size is not provided
+        return call
+
+    return bb.call_te(
+        topi.nn.adaptive_pool,
+        call.args[0],
+        call.attrs.output_size,
+        pool_type="avg",
+        layout=call.attrs.layout,
+    )
+
+
+# def _nn_softmax_cross_entropy(bb: BlockBuilder, call: Call):
+#     def te_softmax_cross_entropy(x, y):
+#         return _te_cross_entropy(topi.nn.softmax(x), y)
+
+#     return bb.call_te(
+#         te_softmax_cross_entropy, args[0], args[1], primfunc_name_hint="softmax_cross_entropy"
+#     )
+
+
+def _sum(bb: BlockBuilder, call: Call):
+    return bb.call_te(topi.sum, call.args[0], call.attrs.axis, call.attrs.keepdims)
+
+
+def _mean(bb: BlockBuilder, call: Call):
+    shape_prod = tvm.tir.const(1, "int32")
+    axis = (
+        call.attrs.axis
+        if call.attrs.axis is not None
+        else range(0, len(call.args[0].struct_info.shape))
+    )
+    for dim in axis:
+        shape_prod = (
+            shape_prod * call.args[0].struct_info.shape[dim if isinstance(dim, int) else dim.value]
+        )
```

Comment:
Negative indexing case will report error. Solution: fix it in Expr getitem https://github.com/tlc-pack/relax/pull/357 or in legalizer https://github.com/mlc-ai/relax/pull/98

Reply:
Yes I left a comment on that side. Could you guys add a test case for Tuple?

Reply:
Sure